readme.md

Welcome to my shitty code! This is meant to sync up Productboard and Gitlab until Productboard inevitably builds their own, much better native integration. By the time you finish reading through this, you'll have all fingers crossed that they integrate ASAP.


**This Code:**
- Pushes new features created in Productboard over to Gitlab when manually triggered via a custom field dropdown in Productboard (Not all PB features want to immediately go to Gitlab, that would be a lot more than will ever be actionable)
- If the custom field dropdown is updated *but* there's already a URL in the custom text field in the Productboard feature, then this script will move the Gitlab issue to the corresponding project from the dropdown values
- Updates any linked issue in Gitlab when it's changed in Productboard (title, description, state/status)
- Updates any linked issue in Productboard when it's changed in Gitlab (title, status)
- When the "push to Gitlab" dropdown's value is changed to "not set" and the "gitlab URL" field has a valid issue URL, deletes the corresponding linked feature from Gitlab


**Requirements/Setup:**
This isn't meant to be a long-term solution (though, I guess many things aren't and look what happens...), so this code is mostly written to be easily maintained, not replicated for other projects/companies/setups. Commenting the hell out of it just in case.


***PRODUCTBOARD***
1. In the "integrations" section, under "Public API," create a new access token, and store it in config.php

2. Create two new custom fields in Productboard:
- "Push to Gitlab" - set as dropdown, and for the values, add your Gitlab project names with IDs in parentheses in the value (e.g., "My cool project (54637)")- these IDs are string matched
- "Gitlab URL" - a text field, this is where the URL to the corresponding issue / product backlog item in Gitlab will go once a successful sync is established (and is how you know it worked)

3. Use the Productboard API to retrieve the IDs[1] of the two custom fields, and put them in their corresponding config variables in config.php
[1] see https://developer.productboard.com/#tag/hierarchyEntitiesCustomFields for instructions

4. Use the Productboard API to create a new subscription[2] to this script, appending /productboard as your route (e.g., https://api.example.com/my/project/path/productboard). This script will respond to the required Productboard validation at this route
[2] see https://developer.productboard.com/#operation/postWebhook for instructions


***GITLAB***
1. Gitlab requires a personal access token for a few key actions (group tokens won't work according to the API docs since an "author" needs to be assigned), so the first step is to log in, and under your own account, head to preferences, then "access tokens", and create a new one with "API" permissions to read/write to the Gitlab API for your project. Add the generated secret to the Gitlab keys array, and enter the Gitlab API URL (latest tested with this code is v4) to config.php

2. Generate a GUID to be passed to this application from Gitlab's webhooks in an X-GITLAB-TOKEN header, and pop that into config.php

3. In the Gitlab GUI, register a webhook with this project's URL, appended by /gitlab (e.g., https://api.example.com/my/project/path/gitlab) at {Group Name} > Settings > Webhooks, with Issue, Note, and Release events, over SSL


**Bringing it all together**
Once you've completed both steps above, you'll need to set the status mappings between Gitlab and Productboard, so that this code can translate between the two.

For Gitlab, the labels field is used, but in conjunction with state (opened/closed). For Productboard, you'll need to use the API to get the API IDs of the statuses you want, and then assign one to open, another to closed, and then map the rest you'd like to use one-to-one between Productboard and Gitlab.

Drop these into the STATUS_MAPPINGS array in config.php. This array does a lot of heavy lifting, so be sure to follow instructions in the comments around it.


That should be it! You're up and running. I've probably forgotten something, but let's be real– hopefully you're not really needing all of this. That'd mean you're setting this up from scratch, and that means this code is being used to do something it was never meant for. Godspeed if that's you.


changelog.md
# Changelog

### [ December 16, 2022 - JP ]
1. Added support for multiple personal access tokens in Gitlab
Now, the owner/creator in Productboard will be reflected in Gitlab, and in the history on an issue. When the owner doesn't match with a configured token, the owner is assigned as the first configured one in this service (currently John Palmer).

2. Syncing of status now considers Gitlab state (opened/closed)
Productboard statuses are now fully mapped, considering whether an issue in Gitlab is open/closed, or using a matching label in the STATUS_MAPPING['statuses'] array. When no matching label is present, the Gitlab state is used.

With the way the mappings are configured currently, this means:
- If Gitlab issue is closed, will show as "closed" in Productboard
- If Gitlab issue is open with no label, will show as "new idea" in Productboard
- If Gitlab issue is open with label, will show as corresponding status in Productboard
- If Gitlab issue is open with multiple labels, will use the first matching label from the configured statuses array

3. Added logging, with options for basic or verbose modes
Outputs to log.txt in the project root.


parts/php/class.gitlab.php
<?

/*

**/

# html -> markdown
require_once __DIR__.'/../../plugins/markdownify/Converter.php';
require_once __DIR__.'/../../plugins/markdownify/ConverterExtra.php';
require_once __DIR__.'/../../plugins/markdownify/Parser.php';


class gitlab {

	public string $generatedUrl, $projectId;


	public function __construct(
		$ownerEmail = ''
	){

		/*
		 Gitlab recommends personal access tokens, so this service will use the one matching the owner of an issue in Productboard. In the event there is not a match or no owner provided, the first item in the array should be used.
		*/
		$this->key = (isset(GITLAB['keys'][$ownerEmail]) && !empty(GITLAB['keys'][$ownerEmail])) ? GITLAB['keys'][$ownerEmail] : array_values(GITLAB['keys'])[0];
		$this->baseUrl = GITLAB['url'];

	}


	private function gl_request( $METHOD, $URL, $CONTENT='' ){

		# To access the Gitlab API, we need to provide a bearer token, authentication:$CFG->['gitlab']['key']
		$context = stream_context_create([
			  'http'=>[
				'method'=>$METHOD,
				'header'=>"Accept-language: en\r\n" .
						  "Authorization: Bearer ".$this->key."\r\n".
						  "Content-type: application/json\r\n",
				'content'=>$CONTENT
			]
		]);

		$response = file_get_contents($URL, false, $context);

//		Gitlab is returning 400 errors but still successfully performing operations, disable this for now
// 		if (str_contains($http_response_header[0], '200')){
// 			return $response;
// 		}
//
// 		return FALSE;

		return $response;

	}


	public function get_issue( $PARAMS ){

		# GET /projects/:id/issues/:issue_iid

		$response = $this->gl_request(
			"GET",
			$this->baseUrl.'/projects/'.$PARAMS->project.'/issues/'.$PARAMS->iid
		);

		if ($response){
			$data = json_decode($response);
			return $data;
		}

		return FALSE;

	}


	public function get_project_id( $PARAMS ){

		# translate a URL-encoded project name into its internal Gitlab ID
		# GET /project_aliases/:name

		$response = $this->gl_request(
			"GET",
			$this->baseUrl.'/projects/'.$PARAMS->path
		);

		if ($response){
			$data = json_decode($response);
			$this->projectId = $data->id;
			return $data;
		}

		return FALSE;

	}


	public function create_issue( $PARAMS ){

		# needs Gitlab admin access

		$MD = new Markdownify\Converter;

		$response = $this->gl_request(
			"POST",
			$this->baseUrl.'/projects/'.$PARAMS->project.'/issues/',
			json_encode([
				'title'			=> $PARAMS->title,
				'description'	=> "[[pb:".$PARAMS->productboardId."]]  \n\n[View in Productboard](".$PARAMS->link.")  \n\n".$MD->parseString($PARAMS->description),
				'labels'		=> $PARAMS->status
			])
		);

		if ($response){
			$data = json_decode($response);
			$this->generatedUrl = $data->web_url;
			return $data;
		}

		return FALSE;

	}


	public function update_issue( $PARAMS ){

		# To prevent circular updates, check for timestamp of last edit of Gitlab issue. If it was updated in the last UPDATE_WAIT seconds, don't update it.
		$issue = $this->get_issue((object)[
			'project'	=> $PARAMS->project,
			'iid'		=> $PARAMS->iid
		]);

		# PUT /projects/:id/issues/:issue_iid
		$MD = new Markdownify\Converter;

		$response = $this->gl_request(
			"PUT",
			$this->baseUrl.'/projects/'.$PARAMS->project.'/issues/'.$PARAMS->iid,
			json_encode([
				'title' 		=> $PARAMS->title,
				'description'	=> "[[pb:".$PARAMS->productboardId."]]  \n\n[View in Productboard](".$PARAMS->link.")  \n\n".$MD->parseString($PARAMS->description),
				'labels'		=> $PARAMS->status
			])
		);

		if ($response){
			$data = json_decode($response);
			$this->generatedUrl = $data->web_url;
			return $data;
		}

		return FALSE;

	}


	public function move_issue( $PARAMS ){

		# /projects/:id/issues/:issue_iid/move?to_project_id={project id}

		$response = $this->gl_request(
			"POST",
			$this->baseUrl.'/projects/'.$PARAMS->fromProject.'/issues/'.$PARAMS->iid.'/move?to_project_id='.$PARAMS->toProject
		);

		if ($response){
			$data = json_decode($response);
			$this->generatedUrl = $data->web_url;
			return $data;
		}

		return FALSE;

	}

	public function delete_issue( $PARAMS ){

		# needs Gitlab admin access
		# DELETE /projects/:id/issues/:issue_iid

		$projectId = $this->get_project_id((object)[
			'path' => $PARAMS->path
		]);

		$response = $this->gl_request(
			"DELETE",
			$this->baseUrl.'/projects/'.$this->projectId.'/issues/'.$PARAMS->iid
		);

		if ($response){
			$data = json_decode($response);
			return $data;
		}

		return FALSE;

	}

}


/* ---------------------------------------*/
// End of file class.gitlab.php
// Location: /gitlab-productboard/parts/php/class.gitlab.php


parts/php/class.logger.php
<?php

/*
This class is a utility to allow writing errors to log.txt,
which is automatically placed in the project root.
*/

class logger{

	public function __construct($PATH){

		return $this->set_file($PATH . '/log.txt');

	}


	public function __sleep(){

		$this->write("\n\n\n\n");
		fclose($this->handle);

	}


	public function __wakeup(){

		$this->set_file($this->path);

	}


	public function set_file( $PATH ){

		if (DEBUG_ON){

			$this->path = $PATH;

			if ($this->handle = fopen($PATH,"a")){

				fwrite($this->handle,"\n- - - - - - - - - - - - - - -\n");
				return $this->handle;

			}

		}

		return FALSE;

	}


	public function write( $STR, $LINE=FALSE, $FILE=FALSE, $FUNCTION=FALSE, $ISVERBOSE=FALSE ){

		global $CFG;

		if (DEBUG_ON){

			$now = date('Y-m-d g:i:sa T');

			$end = '';
			$append = array();
			if ($LINE){ $append[]='Line '.$LINE; }
			if ($FILE){ $append[]=$FILE; }
			if ($FUNCTION){ $append[]=$FUNCTION; }
			if (count($append) > 0){ $end = '('.implode(' * ', $append).")\n"; }

			# Is the line to be written considered verbose? Do we have verbose logging turned on?
			if ($ISVERBOSE && LOG_VERBOSE){

				if (fwrite($this->handle, "[V!] ".$now." :: ".$STR."\n".$end."\n") === FALSE) { return FALSE; }

				return TRUE;

			}else{

				if (fwrite($this->handle, $now." :: ".$STR."\n".$end."\n") === FALSE) { return FALSE; }

				return TRUE;

			}

		}


		return FALSE;

	}


	public function get_file( $PATH='' ){

		if(!empty($PATH)){ $this->path = $PATH; }

		if (!$handle = fopen($this->path, 'a')) {

			 echo "Cannot open logfile.";

			 return FALSE;

		}

		return TRUE;

	}

}


/* ---------------------------------------*/
// End of file class.logger.php
// Location: /gitlab-productboard/parts/php/class.logger.php


parts/php/class.productboard.php
<?

/***
Productboard will send via the POST http request method, and provide a JSON payload that looks something like this:

	{
		 "data":
		 {
			 "eventType": "hierarchy-entity.custom-field-value.updated",
			 "links":
			 {
				 "target": "https://api.productboard.com/hierarchy-entities/custom-fields-values/value?customField.id=898a6004-8aca-490b-aa83-10a4c4c573ac&hierarchyEntity.id=1be161c2-1253-4836-9df1-8c2ec8dddc6e"
			 }
		 }
	 }

	... or this ...

	{
		"data":
		{
			"id": "4639e43a-3851-4566-a506-41ee907af799",
			"eventType": "feature.updated",
			"links":
			{
				"target": "https://api.productboard.com/features/4639e43a-3851-4566-a506-41ee907af799"
			}
		}
	}

**/

require_once __DIR__.'/../../plugins/parsedown/Parsedown.php'; # markdown -> html


class productboard {

	# we'll use the "push to gitlab" and "gitlab URL" custom field IDs from Productboard as the triggers to perform needed actions, which need to be passed into this class on setup.
	public string $event;
	public string $target;
	public object $targetData;

	public function __construct(){

		$this->key = PRODUCTBOARD['key'];
		$this->baseUrl = PRODUCTBOARD['url'];

	}

	private function pb_request( $METHOD, $URL, $CONTENT='' ){

		# To access the Productboard API, we need two header key/value pairs: X-version:1, and authentication:{$this-key}
		$context = stream_context_create([
	  		'http'=>[
				'method'=>$METHOD,
				'header'=>"Accept-language: en\r\n" .
				  		"Authorization: Bearer ".$this->key."\r\n".
						"Content-type: application/json\r\n".
				  		"X-Version: 1\r\n",
				'content'=>$CONTENT
			]
		]);
		// echo "\n\n\ncontent:\n"; # for debugging
		// var_dump($CONTENT);

		$response = file_get_contents($URL, false, $context);
		return $response;

	}

	# Any payload we receive, we probably care about the corresponding links->target's data, so this function pre-emptively retrieves it and loads it into $this->targetData
	public function get_target( $PAYLOAD ){

		$this->event = $PAYLOAD->eventType;			# a string from Productboard telling us the type of event, like "feature.updated"
		$this->target = $PAYLOAD->links->target; 	# this is a direct URL to the Productboard API, corresponding to this object's attributes and values, to be used with get_target()

		$response = $this->pb_request("GET", $this->target);

		if (!empty($response)){
			$obj = json_decode($response);
			$this->targetData = $obj->data;
			return $this->targetData;
		}

		return FALSE;

	}


	# UTILITIES

	# get a custom field value's attributes for a custom field ID on the given entity ID (e.g., a feature ID)
	public function get_field_value_attributes( $FIELDID, $ENTITYID ){

		global $CFG;

		$response = $this->pb_request("GET", $this->baseUrl.'/hierarchy-entities/custom-fields-values/value?customField.id='.$FIELDID.'&hierarchyEntity.id='.$ENTITYID);

		if (!empty($response)){
			$obj = json_decode($response);
			return $obj->data;
		}

		return FALSE;

	}

	# update a custom field value
	public function update_field_value( $PARAMS ){

		global $CFG;

		$response = $this->pb_request(
			"PUT",
			$this->baseUrl.'/hierarchy-entities/custom-fields-values/value?customField.id='.$PARAMS->field.'&hierarchyEntity.id='.$PARAMS->feature,
			json_encode([
				'data' => [
					'type' => 'text',
					'value' => $PARAMS->value
				]
			])
		);

		if (!empty($response)){
			$obj = json_decode($response);
			return $obj->data;
		}

		return FALSE;

	}


	# get a feature's attributes from Productboard
	public function get_feature( $ID ){

		$response = $this->pb_request(
			"GET",
			$this->baseUrl.'/features/'.$ID
		);

		if (!empty($response)){
			$obj = json_decode($response);
			return $obj->data;
		}

		return FALSE;

	}

	# update a feature's attributes in Productboard
	public function update_feature( $PARAMS ){

		# PATCH /features/{id}

		// $PD = new Parsedown();

		$response = $this->pb_request(
			"PATCH",
			$this->baseUrl.'/features/'.$PARAMS->productboardId,
			json_encode([
				'data' => [
					'name'	 		=> $PARAMS->title,

					# replace the [[pb:123-abc]] id stored in the description before updating in Productboard
					# Productboard wants HTML, Gitlab supplies markdown - convert md to html with Parsedown
					# description updating disabled for now, it's screwing with status
					// 'description' 	=> $PD->text(preg_replace('/\[\[.+\]\]/', '', strip_tags($PARAMS->description))),
					'status'		=> [
						'id'		=> $PARAMS->status
					]
				]
			])
		);

		if (!empty($response)){
			$obj = json_decode($response);
			return $obj->data;
		}

		return FALSE;

	}

	# delete a feature in Productboard
	public function delete_feature( $ID ){

		return FALSE;

	}

}


/* ---------------------------------------*/
// End of file class.productboard.php
// Location: /gitlab-productboard/parts/php/class.productboard.php


views/productboard/index.php
<?


/**
 This code is set to receive Productboard events as a webhook, and to take corresponding action in Gitlab with the sent data. Initial JSON payloads from Productboard come with an eventType and a target URL to retrieve data about the affected entity.

 The code below parses data->eventType to drive the action it will take with the data from the corresponding data->links->target.

 For easy debugging, this webhook is called concurrently with one at:
    https://webhook.site/#!/6a79e516-9cbe-46fe-a92e-6b1abdcf57fe
 It receives all the same events and data as this one.
*/


# in config.php, toggles are available to disable the entire sync service or just calls to productboard
if (!ENABLE_SERVICE || !ENABLE_PRODUCTBOARD){

    if (!ENABLE_SERVICE){
        $LOGGER->write('Service was called, but is disabled.', __LINE__, __FILE__);
        echo "Service is disabled.";
    }else{
        $LOGGER->write('Service was called, but Productboard is disabled.', __LINE__, __FILE__);
        echo "Productboard broadcasting is disabled.";
    }
    exit();

}


/**
 When registering a subscription URL to a webhook in Productboard,the registered URL will be pinged to validate that it's expecting pings.

 To validate, we need to return a validationToken that is passed as a query parameter to the registered URL when the subscription is registered.
**/
if (isset($_REQUEST['validationToken'])){

    $LOGGER->write('Webhook was registered with validation token '.$_REQUEST['validationToken'], __LINE__, __FILE__);
	$token = $_REQUEST['validationToken'];
	echo $token;
    exit();

}


/**
 When creating a webhook, Productboard will pass any value you like in an 'authorization' header, and also lets you specify a version number for your webhook, so you can differentiate between requests.
**/
$key = @$_SERVER['authorization'];
$version = @$_SERVER['version']; # not using this right now -JP 09/2022

if (!empty($key) && $key !== PRODUCTBOARD['auth']){

    http_response_code(401);
    $LOGGER->write('Exiting early; authorization header had an invalid value.', __LINE__, __FILE__);
    echo "Please provide a recognized value in the authentication header.";
    exit();

}


/**
 When triggering this webhook, Productboard will act on the returned HTTP response code, and for anything other than a 2XX, will retry the event as many as 8 times with a psuedo-exponentially-scaled sequence of wait times in minutes (1m, 3m, 9m, 27m, 95m... see https://developer.productboard.com/#section/Notification-delivery for full scale)

 Be sure to respond with valid codes so Productboard knows what to do.
*/
if (isset($_SERVER['REQUEST_METHOD']) && strtolower($_SERVER['REQUEST_METHOD']) !== 'post'){
    http_response_code(405);
    $LOGGER->write('Exiting early; request method not allowed. Try POST.', __LINE__, __FILE__);
    echo "Request method not allowed on this resource. Try POST.";
    exit();
}

$LOGGER->write('Parsing body payload...', __LINE__, __FILE__, FALSE, TRUE);
$json = file_get_contents('php://input');
$obj = json_decode($json);
$LOGGER->write('JSON body from payload decoded: '.serialize($obj), __LINE__, __FILE__, FALSE, TRUE);


$LOGGER->write('Grabbing detailed Productboard object attributes...', __LINE__, __FILE__, FALSE, TRUE);
$PB = new productboard();
$PB->get_target($obj->data);
$LOGGER->write('Got target data back from Productboard: '.serialize($PB), __LINE__, __FILE__, FALSE, TRUE);


$httpCode = 200;

$LOGGER->write('Determining which Productboard event to be handled...', __LINE__, __FILE__);

switch($PB->event){

    /***
    Issue in Productboard has been set with a project ID in the "push to Gitlab" field, so:
    A. check URL field of pushed issue to see if Gitlab URL exists for Productboard issue
     - If empty, create a new issue in Gitlab, and populate the URL field when done.
     - If already populated and valid, update issue attributes from Productboard to Gitlab
     - If not empty and not valid, return a friendly error
    **/
    case "hierarchy-entity.custom-field-value.updated":

        // var_dump($PB); #this has all of the data from the originally-sent event, plus some data from the target entity updated
        // exit();

        $LOGGER->write('Custom field value updated, determining type...', __LINE__, __FILE__, FALSE, TRUE);

        $event = (object)[
          'id' => $PB->targetData->hierarchyEntity->id,    # feature ID, a GUID
          'fieldId'=> $PB->targetData->customField->id,    # the ID of the custom field that was changed in Productboard
          'type' => $PB->targetData->hierarchyEntity->type # "feature"
        ];

        /***
        We only want to know when a feature that has this field is updated, *not* if the field configuration itself changes (say, the field is renamed), and we only want to react to changes in the Push to Gitlab URL (otherwise we'll have circular events coming in)
        **/
        if ($event->type=='feature' && $event->fieldId==PRODUCTBOARD['push-to-gitlab']){

            $LOGGER->write('Event type is feature-related, and matches a request to push to gitlab.', __LINE__, __FILE__, FALSE, TRUE);

            # get the feature's attributes & values
            $feature = $PB->get_feature($event->id);

            $LOGGER->write('Prepping to push to Gitlab, got full feature data back from Productboard: '.serialize($feature), __LINE__, __FILE__, FALSE, TRUE);


            # 1. Use the value updated from the dropdown to push or move the issue to Gitlab.
            $pushToGitlabValue = $PB->targetData->value; # when populated, this is an object with a "label" child node, but when no value is set, this is just an empty object (or string?)
            if (!empty($pushToGitlabValue)){

                preg_match("/\([0-9]+\)/", $pushToGitlabValue->label, $matches);
                $gitlabProjectId = str_replace('(','',str_replace(')','',$matches[0]));

                # step back a second, check if this feature has a Gitlab URL already
                $gitlabUrl = $PB->get_field_value_attributes(
                    PRODUCTBOARD['gitlab-url'],    # the custom field ID for the 'Gitlab URL' text field in Productboard
                    $feature->id                   # the hierarchyEntity ID to return the value attributes for
                );

                # if the URL is already set and the dropdown changed (most likely to a new project), we want to move the issue to that corresponding project in Gitlab
                if (!empty($gitlabUrl->value)){

                    $LOGGER->write('Will update in Gitlab, since Gitlab URL has a value: '.$gitlabUrl->value, __LINE__, __FILE__, FALSE, FALSE);

                    # move gitlab issue to the project with $gitlabProjectId specified in update event
                    $matches = [];
                    preg_match('/[^\/]+(\/$|$)/', $gitlabUrl->value, $matches);
                    $gitlabIssueId = $matches[0];
                    $arr = explode('/', str_replace('https://gitlab.com/','',$gitlabUrl->value));
                    $gitlabProjectPath = urlencode($arr[0].'/'.$arr[1]);

                    $GL = new gitlab($feature->owner->email);
                    if ($GL->move_issue((object)[
                        'id'         => $feature->id,       # productboard feature ID
                        'iid'        => $gitlabIssueId,     # gitlab internal issue ID, the last segment of a gitlab web_url
                        'fromProject'=> $gitlabProjectPath, # the urlencoded path or ID/integer for the project the issue is currently in
                        'toProject'  => $gitlabProjectId    # the project ID/int to move the issue to
                    ])){

                        # update returned Gitlab URL in Productboard
                        $PB->update_field_value((object)[
                            'feature'=> $feature->id,
                            'field' => PRODUCTBOARD['gitlab-url'],
                            'value' => $GL->generatedUrl
                        ]);

                    }

                # URL was empty, we're creating a new issue in Gitlab in the specified project, and updating the URL field on the given feature in Productboard
                }else{

                    $LOGGER->write('No Gitlab URL set for Productboard issue, will create a new one in Gitlab.', __LINE__, __FILE__);

                    # convert Productboard status to Gitlab status
                    if ($feature->status->id == STATUS_MAPPING['open']){
                        $updateState    = 'opened';
                        $updateStatus   = '';
                    }elseif ($feature->status->id == STATUS_MAPPING['closed']){
                        $updateState    = 'closed';
                        $updateStatus   = '';
                    }else{
                        # the issue is open, but is somewhere between start and finished, so needs some other status label
                        $updateState    = 'opened';
                        $updateStatus   = STATUS_MAPPING['statuses'][$feature->status->id];
                    }

                    # create new issue in Gitlab
                    $GL = new gitlab($feature->owner->email);
                    if ($GL->create_issue((object)[
                        'project'           => $gitlabProjectId,
                        'title'             => $feature->name,
                        'description'       => $feature->description,
                        'link'              => $feature->links->html,
                        'state'             => $updateState,
                        'status'            => $updateStatus, # hey, if you're tired: no need to consider other labels here, this is a new issue
                        'productboardId'    => $feature->id
                    ])){

                        $LOGGER->write('Issue created in Gitlab, got URL back: '.$GL->generatedUrl, __LINE__, __FILE__, FALSE, TRUE);

                        # update returned Gitlab URL in Productboard
                        $PB->update_field_value((object)[
                            'feature'   => $feature->id,
                            'field'     => PRODUCTBOARD['gitlab-url'],
                            'value'     => $GL->generatedUrl
                        ]);

                        $LOGGER->write('Gitlab URL updated in Productboard for feature ID '.$feature->id, __LINE__, __FILE__);

                    }

                }

            # 2. If the dropdown value has been updated to "not set" (null) in Productboard *and* the Gitlab URL field has a value, then we want to delete it from Gitlab (but not Productboard) based on the Gitlab URL
            }else{

                $LOGGER->write('Dropdown value for Gitlab project has been unset in Productboard, preparing to delete from Gitlab...', __LINE__, __FILE__);

                # step back a second, check if this feature has a Gitlab URL already
                $gitlabUrl = $PB->get_field_value_attributes(
                    PRODUCTBOARD['gitlab-url'],    # the custom field ID for Gitlab URL
                    $PB->targetData->hierarchyEntity->id            # the hierarchyEntity ID to return the value attributes for
                );

                $matches = [];
                preg_match('/[^\/]+(\/$|$)/', $gitlabUrl->value, $matches);
                $gitlabIssueId = $matches[0];
                $arr = explode('/', str_replace('https://gitlab.com/','',$gitlabUrl->value));
                $gitlabProjectPath = urlencode($arr[0].'/'.$arr[1]);


                $LOGGER->write('Set Gitlab URL field in Productboard to empty', __LINE__, __FILE__, FALSE, TRUE);

                # for this feature, set the Gitlab URL field in Productboard to empty
                $PB->update_field_value((object)[
                    'feature'=> $PB->targetData->hierarchyEntity->id,
                    'field' => PRODUCTBOARD['gitlab-url'],
                    'value' => ''
                ]);

                # delete issue from Gitlab
                $GL = new gitlab();
                if (!$GL->delete_issue((object)[
                    'path' => $gitlabProjectPath,
                    'iid'   => $gitlabIssueId
                ])){
                    $LOGGER->write('Issue deleted from Gitlab at '.$gitlabProjectPath.'/issues/'.$gitlabIssueId, __LINE__, __FILE__);
                    $httpCode = 400;
                }

            }

        }
    break;


    /***
     B. A feature was updated in Productboard, and if there is a valid Gitlab URL attached to it, we need to update a corresponding feature in Gitlab.
    **/
    case "feature.updated":

        // var_dump($PB); #this has all of the data from the originally-sent event, plus some data from the target entity updated
        // exit();

        $LOGGER->write('Feature updated in Productboard, updating in Gitlab...', __LINE__, __FILE__);

        # 1. Check if Gitlab URL is set
        $gitlabUrl = $PB->get_field_value_attributes(
            PRODUCTBOARD['gitlab-url'],    # the custom field ID for Gitlab URL
            $PB->targetData->id            # the hierarchyEntity ID to return the value attributes for
        );

        # 2. Get the Gitlab project ID assigned to the feature from the Productboard custom field dropdown value
        $gitlabProjectId = $PB->get_field_value_attributes(
            PRODUCTBOARD['push-to-gitlab'],    # the custom field ID for push to gitlab
            $PB->targetData->id                # the hierarchyEntity ID to return the value attributes for
        );

        # 3. If the Gitlab URL is set, update the title, description, and status in Gitlab
        if (!empty($gitlabUrl->value)){

            $LOGGER->write('Gitlab URL value is set, proceeding with update.', __LINE__, __FILE__, FALSE, TRUE);

            preg_match('/\([0-9]+\)/', $gitlabProjectId->value->label, $projectId);
            preg_match('/\/[0-9]+/', $gitlabUrl->value, $urlParts);

            $pid = str_replace('(', '', str_replace(')','',$projectId[0]));
            $iid = str_replace('(', '', str_replace(')','', str_replace('/','',$urlParts[0])));

            $GL = new gitlab($PB->targetData->owner->email);
            $issue = $GL->get_issue((object)[
               'project' => $pid,
               'iid'    =>  $iid
            ]);

            $LOGGER->write('Got existing issue to be updated from Gitlab, extracting updated_at and labels...'.serialize($issue), __LINE__, __FILE__, FALSE, TRUE);

            # filter out any status labels from the existing labels, to avoid conflicts
            $filteredLabels = array_filter($issue->labels, function($item) {
                return !in_array($item, array_values(STATUS_MAPPING['statuses']));
            });
            $existingLabels = (empty($issue->labels)) ? '' : implode(',', $filteredLabels);

            $LOGGER->write('Got existing labels, if any:'.$existingLabels, __LINE__, __FILE__, FALSE, TRUE);


            # To prevent circular updates, check for timestamp of last edit of Gitlab issue. If it was updated in the last 30 seconds, don't update it.
            $diffSeconds = (new DateTimeImmutable("now"))->format('U')-(new DateTimeImmutable($issue->updated_at))->format('U');

            if ($diffSeconds > UPDATE_WAIT){

                $LOGGER->write('Wait threshold passed with '.$diffSeconds.' seconds, updating this feature in Gitlab. Timing was:'."\nnow: ".(new DateTimeImmutable("now"))->format('c').' / '.(new DateTimeImmutable("now"))->format('U')."\nupdatedAt: ".(new DateTimeImmutable($issue->updated_at))->format('c').' / '.(new DateTimeImmutable($issue->updated_at))->format('U'), __LINE__, __FILE__, FALSE, TRUE);


                # convert Productboard status to Gitlab status
                if ($PB->targetData->status->id == STATUS_MAPPING['open']){

                    # the Productboard status ID matches the 'open' state mapping to Gitlab
                    $updateState    = 'opened';
                    $updateStatus   = $existingLabels;

                }elseif ($PB->targetData->status->id == STATUS_MAPPING['closed']){

                    # the Productboard status ID matches the 'closed' state mapping to Gitlab
                    $updateState    = 'closed';
                    $updateStatus   = $existingLabels;

                }else{

                    # the issue is open, but is somewhere between start and finished, so needs some other status label
                    $updateState    = 'opened';
                    $updateStatus   = STATUS_MAPPING['statuses'][$PB->targetData->status->id].','.$existingLabels;

                }

                $feature = (object)[
                  'project'         => $pid,
                  'productboardId'  => $PB->targetData->id,
                  'iid'             => $iid,
                  'title'           => $PB->targetData->name,
                  'description'     => $PB->targetData->description, # productboard delivers HTML, gitlab wants markdown
                  'link'            => $PB->targetData->links->html,
                  'state'           => $updateState,
                  'status'          => $updateStatus
                ];

                $LOGGER->write('Data ready to update in Gitlab: '.serialize($feature), __LINE__, __FILE__, FALSE, TRUE);

                if (!$GL->update_issue($feature)){
                    $LOGGER->write('Issue failed to update in Gitlab.', __LINE__, __FILE__);
                    $httpCode = '400';
                }

            }else{

                echo "No need to update, we're under the wait time threshold and prevented a circular callback.";

                $LOGGER->write('No need to update in Gitlab, we\'re under the wait time threshold and prevented a circular callback.'."\n".'Difference in seconds:'.$diffSeconds.' based on:'."\n".'now:'.(new DateTimeImmutable("now"))->format('c').' / '.(new DateTimeImmutable("now"))->format('U')."\n".'updatedAt:'.$issue->updated_at.' / '.(new DateTimeImmutable($issue->updated_at))->format('U'), __LINE__, __FILE__);

            }

        }else{

            # 3. If the Gitlab URL is not set, do nothing
            echo "No Gitlab URL is set, no target issue in Gitlab to update.";

            $LOGGER->write('No Gitlab URL is set, no target issue in Gitlab to update.', __LINE__, __FILE__);

        }

    break;



    default:
        # No recognized event was sent, do nothing
        $LOGGER->write('No recognized event type was sent from Productboard (this is fine, this webhook is registered for many events), ignoring.', __LINE__, __FILE__, FALSE, TRUE);
    break;

}

http_response_code($httpCode);


/* ---------------------------------------*/
// End of file index.php
// Location: /gitlab-productboard/views/productboard/index.php
